feat: handle vitest workspace configuration

codepunkt opened the pull request:
The `vitest` plugin is not properly handling [workspace configurations](https://vitest.dev/guide/workspace.html).

This PR should enable that, and thus detect various things (e.g. `globalSetup`, `setupFiles`, `reporters` or `coverage.provider`) inside a `workspace` array - like in this example:

```ts
import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    workspace: [
      {
        test: {
          name: 'integration',
          setupFiles: './vitest.integration.setup.mjs',
        },
      },
      {
        test: {
          globalSetup: './vitest.unit.setup.ts',
          name: 'unit',
        },
      },
    ],
  },
});
```

<!--

- Try to author code and/or docs similar to the rest of the repository
- Run `npm run format` (from root)
- Run `npm test` (from root)

Through [the CI workflow][1] the changes will be tested in Ubuntu, macOS and Windows. The changes will also be [tested
against a number of external projects][2].

[1]: https://github.com/webpro-nl/knip/blob/main/.github/workflows/ci.yml
[2]: https://github.com/webpro-nl/knip/blob/main/.github/workflows/integration.yml

-->

webpro commented on packages/knip/test/plugins/vitest8.test.ts:
```diff
@@ -0,0 +1,26 @@
+  });
+
+  assert(issues.unresolved['vitest.config.ts']['./vitest.integration.setup.mjs']);
+  assert(issues.unresolved['vitest.config.ts']['./vitest.unit.setup.ts']);
```
Thanks for the PR! Happy to merge. If you could please `touch` those (empty) files there? Then this test file and the `counters` are tidy too 🍃

codepunkt replied:
https://github.com/webpro-nl/knip/pull/1003/commits/c2ce2934a6a907e3ffef85d83e147b373b743019

codepunkt replied:
I'm not 100% sure that I understand the counters and issues returned by `await main()`. When i `touch` the files, the test fails because they're not unresolved then.

How exactly would `touch`ing the files make this test file and `counters` tidy?

webpro replied:
Sorry, I wasn't clear. With "tidy" I meant that the issue counters end up being zero (`0`). When the files are there (i.e. not unresolved), there are no issues left (only a number of processed/total files). Then we don't need to assert numbers for `unresolved` etc. which in my mind is "tidy".

webpro commented on packages/knip/src/plugins/vitest/index.ts:
```diff
@@ -45,7 +45,19 @@ const findConfigDependencies = (localConfig: ViteConfig, options: PluginOptions)
 
-  return [...[...environments, ...reporters, ...coverage].map(id => toDependency(id)), ...setupFiles, ...globalSetup];
+  const workspaceDependencies: Input[] = [];
+  if (testConfig.workspace !== undefined) {
```
Please update the types for this plugin (https://github.com/webpro-nl/knip/pull/1003/checks#step:6:18)

codepunkt replied:
https://github.com/webpro-nl/knip/pull/1003/commits/27564a4a14a97a7103d604ff21e36fa9a8d1324c

webpro commented:
Thanks! Only a few lint issues left it seems.

webpro commented:
Here are the commands that can be used for QA and that should go green: https://github.com/webpro-nl/knip/blob/main/.github/DEVELOPMENT.md#qa

There's currently one job in Run Knip against external projects that fails, we can ignore that.

codepunkt commented:
`bun knip:production` wasn't found.
`bun lint` had an error due to unused variables and a small spacing issue, which I've fixed.

`bun run test` failed on the new testcase that I've added, because I'm specifically testing the non-existance of files, but you asked me to add them. I've sinced removed them, and now everything listed in the QA link you've sent (except `knip:production`) should pass.

Is there anything else I can help with?

webpro commented:
> `bun knip:production` wasn't found. 

Forgot to update that one, should be `bun knip --strict` now.

codepunkt commented:
My thanks go to you! 👍🏻

webpro commented:
:rocket: _This pull request is included in v5.47.0. See [Release 5.47.0](https://github.com/webpro-nl/knip/releases/tag/5.47.0) for release notes._

_Using Knip in a commercial project? Please consider [becoming a sponsor](https://knip.dev/sponsors)._

fshowalter commented:
This doesn’t account for `extends` so if you define your coverage engine as Istanbul in the root config and set `extends: true` in the workspace, Knip will think the workspaces have no  coverage definition and default to v8, leading Knip to flag it as a missing dep.
